src/gfn/estimators.py

```diff
@@ -561,124 +706,82 @@ def to_probability_distribution(
+        logits[Ga.NODE_INDEX_KEY][~masks[Ga.NODE_INDEX_KEY]] = -float("inf")
+        logits[Ga.EDGE_CLASS_KEY][~masks[Ga.EDGE_CLASS_KEY]] = -float("inf")
+        logits[Ga.EDGE_INDEX_KEY][~masks[Ga.EDGE_INDEX_KEY]] = -float("inf")
+
+        # The following operations are to ensure that the logits are valid, i.e.,
+        # contain at least one finite entry.
 
         # Check if no possible edge can be added,
         # and assert that action type cannot be ADD_EDGE
-        no_possible_edge_index = torch.isneginf(logits[GraphActions.EDGE_INDEX_KEY]).all(
-            -1
-        )
+        no_possible_edge_index = torch.isneginf(logits[Ga.EDGE_INDEX_KEY]).all(-1)
         assert torch.isneginf(
-            logits[GraphActions.ACTION_TYPE_KEY][
-                no_possible_edge_index, GraphActionType.ADD_EDGE
-            ]
+            logits[Ga.ACTION_TYPE_KEY][no_possible_edge_index, GaType.ADD_EDGE]
         ).all()
-        logits[GraphActions.EDGE_INDEX_KEY][no_possible_edge_index] = 0.0
+        logits[Ga.EDGE_INDEX_KEY][no_possible_edge_index] = 0.0
 
         # Check if no possible edge class can be added,
         # and assert that action type cannot be ADD_EDGE
-        no_possible_edge_class = torch.isneginf(logits[GraphActions.EDGE_CLASS_KEY]).all(
-            -1
-        )
-        assert torch.isneginf(
-            logits[GraphActions.ACTION_TYPE_KEY][
-                no_possible_edge_class, GraphActionType.ADD_EDGE
-            ]
-        ).all()
-        logits[GraphActions.EDGE_CLASS_KEY][no_possible_edge_class] = 0.0
-
-        # Check if no possible node can be added,
-        # and assert that action type cannot be ADD_NODE
-        no_possible_node_class = torch.isneginf(logits[GraphActions.NODE_CLASS_KEY]).all(
-            -1
-        )
-        no_possible_node_index = torch.isneginf(logits[GraphActions.NODE_INDEX_KEY]).all(
-            -1
-        )
+        no_possible_edge_class = torch.isneginf(logits[Ga.EDGE_CLASS_KEY]).all(-1)
         assert torch.isneginf(
-            logits[GraphActions.ACTION_TYPE_KEY][
-                no_possible_node_class & no_possible_node_index, GraphActionType.ADD_NODE
-            ]
+            logits[Ga.ACTION_TYPE_KEY][no_possible_edge_class, GaType.ADD_EDGE]
         ).all()
-        logits[GraphActions.NODE_CLASS_KEY][no_possible_node_class] = 0.0
-        logits[GraphActions.NODE_INDEX_KEY][no_possible_node_index] = 0.0
-
-        probs = {}
+        logits[Ga.EDGE_CLASS_KEY][no_possible_edge_class] = 0.0
+
+        # Check if no possible node can be added; if either class OR index has no
+        # valid options, disallow ADD_NODE in action type and set harmless finite
+        # fallbacks for the unused components.
+        no_possible_node_class = torch.isneginf(logits[Ga.NODE_CLASS_KEY]).all(-1)
+        no_possible_node_index = torch.isneginf(logits[Ga.NODE_INDEX_KEY]).all(-1)
+        no_possible_add_node = no_possible_node_class | no_possible_node_index
```

Comment:
we really need to fix this first, because I think it would make the add action impossible (like triangle tutorial giving 0%)

Or we temporarly revert to the previous workflow that converts to probability